sys/kern/kftrace.c

```diff
@@ -0,0 +1,99 @@
+#include <sys/kftrace.h>
+#include <sys/thread.h>
+#include <sys/types.h>
+#include <sys/malloc.h>
+
+/* Kernel Function Trace
+ *
+ * In this file there are defined structures and functions that are used to
+ * implement function tracking in kernel.
+ *
+ * When entering or exiting a function an event is added to `kft_event_list`.
+ * The size of the list is specified by `KFT_EVENT_MAX`.
+ *
+ * When the list is full the function `kft_flush` is invoked. Currently it only
+ * resets the event count. The main purpose of it is to set a breakpoint on it
+ * and dump the contents of the list of events to a file by the debugger.
+ *
+ * Information about the event is compressed into a single 64 bit value.
+ * Below is a picture of layout of information encoded in that value.
+ *
+ * +---------+-----------+-----------+------------+
+ * | 24 bits | 31 bits   |   8 bits  |   1 bit    |
+ * |    PC   | timestamp | thread id | event type |
+ * +---------+-----------+-----------+------------+
+ */
+
+typedef uint64_t kft_event_t;
+
+#ifdef MIPS_KFT
+#define KFT_EVENT_MAX 8000
+#else
+#define KFT_EVENT_MAX 1000000
+#endif
+
+static kft_event_t *kft_event_list;
+static unsigned kft_used = 0;
+static bool kft_enabled = false;
+
+#define PC_MASK 0xFFFFFF          /* 24 bits */
+#define TIMESTAMP_MASK 0x7FFFFFFF /* 31 bits */
+#define THREAD_MASK 0xF           /* 8 bits */
+#define TYPE_MASK 0x1             /* 1 bit */
+
+#define PC_SHIFT 40
+#define TIMESTAMP_SHIFT 9
+#define THREAD_SHIFT 1
+#define TYPE_SHIFT 0
+
+typedef enum kft_event_type {
+  KFT_EVENT_FUN_IN,
+  KFT_EVENT_FUN_OUT,
+} kft_event_type_t;
+
+KMALLOC_DEFINE(M_KFT, "kft events buffer");
+
+void init_kftrace(void) {
+  kft_event_list =
+    kmalloc(M_KFT, sizeof(kft_event_t) * (KFT_EVENT_MAX + 5), M_WAITOK);
+  kft_enabled = true;
```

Comment:
Why introduce `kft_enabled` if `kft_event_list` already has the information we need?

Reply:
Right, removed